src/methods.rs

```diff
@@ -92,6 +99,64 @@ impl LateLintPass for MethodsPass {
+                    if inner_name.node.as_str() == "map"
+                            && match_type(cx, cx.tcx.expr_ty(&inner_args[0]), &OPTION_PATH) {
+                        // lint message
+                        let msg =
+                            "called `map(f).unwrap_or(a)` on an Option value. This can be done \
+                             more directly by calling `map_or(a, f)` instead";
+                        // get args to map() and unwrap_or()
+                        let map_arg = snippet(cx, inner_args[1].span, "..");
+                        let unwrap_arg = snippet(cx, args[1].span, "..");
+                        // lint, with note if neither arg is > 1 line and both map() and
+                        // unwrap_or() have the same span
+                        if map_arg.lines().count() > 1 || unwrap_arg.lines().count() > 1
+                                || inner_args[1].span.expn_id != args[1].span.expn_id {
+                            span_lint(cx, OPTION_MAP_UNWRAP_OR, expr.span, msg);
+                        }
+                        else {
+                            let span = mk_sp(inner_name.span.lo, name.span.hi);
+                            span_note_and_lint(
+                                cx, OPTION_MAP_UNWRAP_OR, expr.span, msg, span,
+                                &format!("replace this with map_or({1}, {0})",
+                                         map_arg, unwrap_arg)
+                            );
+                        };
+                    }
+                }
+            }
+            // check Option.map(_).unwrap_or_else(_)
+            else if name.node.as_str() == "unwrap_or_else" {
+                if let ExprMethodCall(ref inner_name, _, ref inner_args) = args[0].node {
+                    if inner_name.node.as_str() == "map"
+                            && match_type(cx, cx.tcx.expr_ty(&inner_args[0]), &OPTION_PATH) {
+                        // lint message
+                        let msg =
+                            "called `map(f).unwrap_or_else(g)` on an Option value. This can be \
+                             done more directly by calling `map_or_else(g, f)` instead";
+                        // get args to map() and unwrap_or_else()
+                        let map_arg = snippet(cx, inner_args[1].span, "..");
+                        let unwrap_arg = snippet(cx, args[1].span, "..");
+                        // lint, with note if neither arg is > 1 line and both map() and
+                        // unwrap_or_else() have the same span
+                        if map_arg.lines().count() > 1 || unwrap_arg.lines().count() > 1
+                                || inner_args[1].span.expn_id != args[1].span.expn_id {
+                            span_lint(cx, OPTION_MAP_UNWRAP_OR_ELSE, expr.span, msg);
+                        }
+                        else {
+                            let span = mk_sp(inner_name.span.lo, name.span.hi);
```

Comment:
Actually in case of a macro just using the regular span shows a  macro backtrace, however, `mk_sp` can end up with very strange results and we should never use it.

It's okay to lint within a macro for this lint since there's a  good chance that such patterns would be found in the macro itself, rather than on a specific expansion.

(We turn off some lints in macros because they deal with cases which can happen in specific macro expansions, i.e. `#[derive(Foo)] struct Bar;` is likely to have a lot of useless `if` or `match` statements inside the expansion, but these aren't useless in the general case.)